src/navigation/_CollapsingNavigationMenu.jsx

```diff
@@ -86,106 +69,233 @@ const openKeysToItem = (menuItems, selectedPath) => keysToItem(menuItems, select
+      cursor.current = visibleNodes.findIndex((el) => el.id === currentNodeId.current);
     }
-    // If the current item is not visible, scroll the item into view.
-    if (currentItemPosition.bottom > navigationMenuPosition.bottom || currentItemPosition.top < navigationMenuPosition.top) {
-      selectedItem.current.scrollIntoView();
+    if (previousSelectedPath.current !== selectedPath) {
+      const selectedItemPosition = selectedItem?.current ? selectedItem.current.getBoundingClientRect() : null;
+      const navigationMenuPosition = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect();
+      if (selectedItemPosition && navigationMenuPosition && (selectedItemPosition.bottom > navigationMenuPosition.bottom || selectedItemPosition.top < navigationMenuPosition.top)) {
+        selectedItem.current.scrollIntoView();
+      }
+      previousSelectedPath.current = selectedPath;
     }
-  });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [openKeys]);
+
+  /**
+   * Ensures the parent of the currently selected item is expanded.
+   */
+  useEffect(() => {
+    setOpenKeys({ ...openKeys, ...openKeysToItem(menuItems[0], selectedPath) });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [selectedPath]);
 
   const handleOnClick = (event, item) => {
+    // Don't update state if function is called through handleKeyDown
+    if (event.type === 'click') {
+      const eventTargetId = event.target.getAttribute('id');
+      currentNodeId.current = eventTargetId;
+      cursor.current = visibleNodes.findIndex((el) => el.id === eventTargetId);
+      focusCurrentNode();
+    }
     if (!item.childItems) {
       onSelect(item.path);
       return;
     }
     setOpenKeys({ ...openKeys, [item.path]: !openKeys[item.path] });
   };
 
+  const handleDownArrow = () => {
+    if (cursor.current + 1 < visibleNodes.length) {
+      cursor.current += 1;
+      currentNodeId.current = visibleNodes[cursor.current].id;
+      focusCurrentNode();
+    }
+  };
+
+  const handleUpArrow = () => {
+    if (cursor.current >= 1) {
+      cursor.current -= 1;
+      currentNodeId.current = visibleNodes[cursor.current].id;
+      focusCurrentNode();
+    }
+  };
+
+  /**
+   * Finds parent of the current node, used for left arrow functionality
+   */
+  const findParentNode = () => {
+    if (!currentNodeId.current) {
+      return;
+    }
+    const parentId = visibleNodes.find((el) => el.id === currentNodeId.current).parent;
+    if (parentId !== '') {
+      cursor.current = visibleNodes.findIndex((el) => el.id === parentId);
+      currentNodeId.current = visibleNodes[cursor.current].id;
+      focusCurrentNode();
+    }
+  };
+
+  /**
+   * Finds the first node starting with the given character.
+   * Starts at the cursors current position, and wraps around to the beginning of the menu if a match isn't found.
+   * @param {*} char The character to search by
+   */
+  const findNodeMatching = (char) => {
```

Comment:
This is probably a question for UX, but should we always find the first matching item starting at index 0, or always the next item starting at the current index for the menu item that starts with the letter?

Focusing a node on the secondary nav test and then press the letter `f` on the keyboard will iterate the cursor through the entire list one by one:

![image](https://user-images.githubusercontent.com/6720991/88885806-cf6ac200-d1fe-11ea-9e87-07f359ef04f2.png)

Reply:
The reason I decided to start the search from the cursor is because it is the behavior described by the WAI-ARIA specifications for a navigatable treeview.